### DISCUSSION TITLE
DatabaseManager IsInstalled may need to be changed

### DISCUSSION THREAD

**Discussion by jucchytil** [oqtane/oqtane.framework#1234] [Ideas]
Sorry for not taking the time to create a pull request.

It seems that it more appropriate to consider IsInstalled to be successful if it can get a tenant.

![image](https://user-images.githubusercontent.com/17008787/114284857-689dd980-9a18-11eb-9e9a-0bed046233dc.png)

Your thoughts?

Would it break anything?

Joe

**Comment by sbwalker** [comment]
Result.success is initialized to false in the first line of the method. Are you encountering an issue with the behavior of this code?

**Comment by jucchytil** [comment]
Yes.

I'm setting up an Azure CI/CD pipeline with your source code coming from my repo.
I create the database beforehand so I have a connection string that I can put in my WebApp configuration.
When line 42 finds the connection string it tries to see if a tenant exists in line 52. When it fails due to the Tenant table not existing, the result is returned with a message Master Database Not Installed Correctly. But because the result was set to success in line 44, it never tries to create the database later on.

Also, I've summitted a suggestion to modify the configuration setup to add environment variables so they can be populated from Azure WebApp configuration.

**Reply by sbwalker** [reply]
Oqtane supports 2 installation modes - interactive and non-interactive - which are controlled by the Installation node in appsettings.json. 

If the Installation node values are unspecified, then the framework assumes the installation is interactive. It then checks the ConnectionString, and if the value is unspecified it shows the Install Wizard which allows a user to proceed with the installation. If the ConnectionString is specified it assumes the framework was installed previously and it then checks if it can connect to the database and if the database is in the correct format. If so, it launches the framework, and if not, it displays an error message. Note that it used to display the Install Wizard rather than display an error message - however that exposed a potential security issue, as it could result in the Install Wizard being displayed to a normal site visitor if the database became unreachable due to some intermittent network or infrastructure issue. So if you are installing on Azure and are creating the database in advance and want to use interactive mode, then do not specify the ConnectionString - leave it blank.

If the Installation node values are specified, then framework assumes the installation is non-interactive. In this mode it expects the ConnectionString to be specified in advance. It then validates the installation parameters specified and if they are correct, it then then checks if it can connect to the database and if the database is in the correct format. If not, it attempts to create the database and provision the schema. So if you are installing on Azure and are creating the database in advance and want to use non-interactive mode, then you need to specify the Installation node settings and specify the ConnectionString.

There are MANY critical use cases which run through DatabaseManager. A few of them include:

- clean interactive initial installation
- clean non-interactive initial installation
- interactive addition of a new tenant in an existing installation
- non-interactive upgrade of all tenants in an existing installation

Any modification to this logic has to be done with great care to ensure there are no regression issues for any use case.

( btw... you might want to edit your reply above and remove the email reply details )
